File: leetcode/0283. Move Zeroes/Solution.swift
// https://leetcode.com/problems/move-zeroes/

class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        var lastNonZeroFoundAt = 0, current = 0
        while current < nums.count {
            if nums[current] != 0 {
                nums.swapAt(lastNonZeroFoundAt, current)
                lastNonZeroFoundAt += 1
            }

            current += 1
        }
    }
}

let s = Solution()
var nums = [0, 1, 0, 3, 12]
s.moveZeroes(&nums)
assert(nums == [1, 3, 12, 0, 0])

File: leetcode/0283. Move Zeroes/Solution.py
# https://leetcode.com/problems/move-zeroes/

from typing import List

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        iteratingIndex = 0
        nextNonZeroPlacementIndex = 0
        while iteratingIndex < len(nums):
            if nums[iteratingIndex]:
                nums[nextNonZeroPlacementIndex], nums[iteratingIndex] = nums[iteratingIndex], nums[nextNonZeroPlacementIndex]
                nextNonZeroPlacementIndex += 1
            
            iteratingIndex += 1

s = Solution()
nums = [0, 1, 0, 3, 12]
s.moveZeroes(nums)
assert nums == [1, 3, 12, 0, 0]
